- Fix RocksDB SIGILL error on Raspberry PI 4
  https://github.com/facebook/rocksdb/commit/29f7bbef995bdf83098963799c66af742e95373f
- Fix compilation on Apple Silicon
  https://github.com/facebook/rocksdb/commit/ee4bd4780b321ddb5f92a0f4eb956f2a2ebd60dc
- Fix build on FreeBSD/powerpc64(le)
  https://github.com/facebook/rocksdb/commit/66e54c59847e316bb78dce786d2206d500bf8a90
- Add elf_aux_info() support for OpenBSD
  https://github.com/facebook/rocksdb/pull/12967

https://jira.mariadb.org/browse/MDEV-29875

Index: storage/rocksdb/rocksdb/util/crc32c.cc
--- storage/rocksdb/rocksdb/util/crc32c.cc.orig
+++ storage/rocksdb/rocksdb/util/crc32c.cc
@@ -35,10 +35,20 @@
 #define AT_HWCAP2 26
 #endif
 
+#elif __FreeBSD__ && ROCKSDB_AUXV_ELF_AUX_INFO_PRESENT
+#include <machine/cpu.h>
+#include <sys/auxv.h>
+#include <sys/elf_common.h>
+#elif __OpenBSD__ && ROCKSDB_AUXV_ELF_AUX_INFO_PRESENT
+#include <sys/auxv.h>
 #endif /* __linux__ */
 
 #endif
 
+#if defined(HAVE_ARM64_CRC)
+bool pmull_runtime_flag = false;
+#endif
+
 namespace ROCKSDB_NAMESPACE {
 namespace crc32c {
 
@@ -457,6 +467,18 @@ static int arch_ppc_probe(void) {
 
   return arch_ppc_crc32;
 }
+#elif __FreeBSD__ || __OpenBSD__
+static int arch_ppc_probe(void) {
+  unsigned long cpufeatures;
+  arch_ppc_crc32 = 0;
+
+#if defined(__powerpc64__) && defined(ROCKSDB_AUXV_ELF_AUX_INFO_PRESENT)
+  elf_aux_info(AT_HWCAP2, &cpufeatures, sizeof(cpufeatures));
+  if (cpufeatures & PPC_FEATURE2_HAS_VEC_CRYPTO) arch_ppc_crc32 = 1;
+#endif  /* __powerpc64__ */
+
+  return arch_ppc_crc32;
+}
 #endif  // __linux__
 
 static bool isAltiVec() {
@@ -468,7 +490,7 @@ static bool isAltiVec() {
 }
 #endif
 
-#if defined(__linux__) && defined(HAVE_ARM64_CRC)
+#if defined(HAVE_ARM64_CRC)
 uint32_t ExtendARMImpl(uint32_t crc, const char *buf, size_t size) {
   return crc32c_arm64(crc, (const unsigned char *)buf, size);
 }
@@ -488,10 +510,11 @@ std::string IsFastCrc32Supported() {
   has_fast_crc = false;
   arch = "PPC";
 #endif
-#elif defined(__linux__) && defined(HAVE_ARM64_CRC)
+#elif defined(HAVE_ARM64_CRC)
   if (crc32c_runtime_check()) {
     has_fast_crc = true;
     arch = "Arm64";
+    pmull_runtime_flag = crc32c_pmull_runtime_check();
   } else {
     has_fast_crc = false;
     arch = "Arm64";
@@ -1220,8 +1243,9 @@ uint32_t crc32c_3way(uint32_t crc, const char* buf, si
 static inline Function Choose_Extend() {
 #ifdef HAVE_POWER8
   return isAltiVec() ? ExtendPPCImpl : ExtendImpl<Slow_CRC32>;
-#elif defined(__linux__) && defined(HAVE_ARM64_CRC)
+#elif defined(HAVE_ARM64_CRC)
   if(crc32c_runtime_check()) {
+    pmull_runtime_flag = crc32c_pmull_runtime_check();
     return ExtendARMImpl;
   } else {
     return ExtendImpl<Slow_CRC32>;
